soaktest/soakclient_transaction/soakclient_transaction.go

```diff
@@ -0,0 +1,495 @@
+	partitionPositionMap := make(map[int32]kafka.TopicPartition)
+	ticker := time.NewTicker(1000 * time.Millisecond)
+
+	c, err := kafka.NewConsumer(&kafka.ConfigMap{
+		"bootstrap.servers":        broker,
+		"group.id":                 groupId,
+		"go.events.channel.enable": true,
+		"enable.auto.commit":       false,
+		"statistics.interval.ms":   50,
+		"sasl.mechanisms":          "PLAIN",
+		"security.protocol":        "SASL_SSL",
+		"sasl.username":            ccloudAPIKey,
+		"sasl.password":            ccloudAPISecret})
+
+	if err != nil {
+		fmt.Printf("Failed to create consumer: %s\n", err)
+	}
+
+	err = c.Subscribe(inputTopic, nil)
+	if err != nil {
+		fmt.Printf("Failed to subscribe to topic: %s with error %s\n",
+			inputTopic, err)
+		return terminate(err, *c)
+	}
+
+	p, err := kafka.NewProducer(&kafka.ConfigMap{
+		"bootstrap.servers":      broker,
+		"transactional.id":       transactionId,
+		"statistics.interval.ms": 50,
+		"sasl.mechanisms":        "PLAIN",
+		"security.protocol":      "SASL_SSL",
+		"sasl.username":          ccloudAPIKey,
+		"sasl.password":          ccloudAPISecret,
+	})
+	if err != nil {
+		fmt.Errorf("Failed to create transactional producer: %s", err)
+		return terminate(err, *c)
+	}
+
+	maxDuration, err := time.ParseDuration("15s")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%s", err)
+		return terminate(err, *c)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), maxDuration)
+	defer cancel()
+
+	// If InitTransactions succeed, will continue to BeginTransaction
+	// If InitTransactions failed with err.(kafka.Error).IsRetriable(),
+	// sleep SLEEP_TIME then retry
+	// If failed with other errors, return the fatal error
+	for i := 0; i < RETRY_NUM; i++ {
+		err = p.InitTransactions(ctx)
+		if err != nil {
+			fmt.Errorf("InitTransactions() failed: %s\n", err)
+			if err.(kafka.Error).IsRetriable() {
+				if i == RETRY_NUM-1 {
+					fmt.Errorf("InitTransactions() failed after %d "+
+						"times retries: %s\n", RETRY_NUM, err)
+					return terminate(err, *c)
+				}
+				time.Sleep(SLEEP_TIME)
+				continue
+			} else {
+				fmt.Errorf("InitTransactions() failed: %s\n", err)
+				return terminate(err, *c)
+			}
+		} else {
+			break
+		}
+	}
+
+	//Start producer transaction.
+	err = p.BeginTransaction()
+	if err != nil {
+		fmt.Errorf("Failed to begin transaction %s\n", err)
+		return terminate(err, *c)
+	}
+
+	var msgid uint64
+	run := true
+	commited := false
+	for run == true {
+		select {
+		case sig := <-sigchan:
+			fmt.Printf("Caught signal %v: terminating\n", sig)
+			run = false
+			if msgid != 0 && !commited {
+				transactionCommitNum, err = commitTransaction(
+					getAllPartitionPositions(partitionPositionMap),
+					*p,
+					*c,
+					transactionCommitNum)
+			}
+			soaktest.PrintConsumerStatus()
+			return err
+		case <-ticker.C:
+			if msgid != 0 && !commited {
+				transactionCommitNum, err = commitTransaction(
+					getAllPartitionPositions(partitionPositionMap),
+					*p,
```

Comment:
We use pointers for the client instances in all APIs so I think you can just pass them as pointers and not dereference them